zingolib/src/wallet/keys/ledger.rs

```diff
@@ -0,0 +1,802 @@
+        if !desired_receivers.transparent || !desired_receivers.sapling {
+            return Err("The wallet is not capable of producing desired receivers.".to_string());
+        }
+
+        let previous_num_addresses = self.addresses.len();
+
+        if desired_receivers.transparent {
+            let count = self
+                .transparent_addrs
+                .read()
+                .map_err(|e| format!("Error: {e}"))?
+                .len();
+
+            let path = Self::t_derivation_path(config.get_coin_type(), count as _);
+
+            let t_pubkey = futures::executor::block_on(self.get_t_pubkey(&path))
+                .map_err(|e| format!("Error: {e}"))?;
+
+            let mut path_u32 = [0u32; 5];
+            path.into_iter()
+                .map(|idx| idx.index())
+                .zip(path_u32.iter_mut())
+                .for_each(|(a, b)| *b = a);
+
+            // This is deprecated. Not sure what the alternative is,
+            // other than implementing it ourselves.
+            let t_addrs = zcash_primitives::legacy::keys::pubkey_to_address(&t_pubkey);
+            let ua = UnifiedAddress::from_receivers(None, None, Some(t_addrs));
+
+            if self
+                .addresses_write_lock
+                .swap(true, atomic::Ordering::Acquire)
+            {
+                return Err("addresses_write_lock collision!".to_string());
+            }
+            let ua = match ua {
+                Some(address) => address,
+                None => {
+                    self.addresses_write_lock
+                        .swap(false, atomic::Ordering::Release);
+                    return Err(
+                        "Invalid receivers requested! At least one of sapling or transparent required, orchard is not support"
+                            .to_string(),
+                    );
+                }
+            };
+
+            // TODO: Do we want to keep to separate lists for transparent address and keys?
+            // what is appending in one fails whereas the other succeded?
+            self.addresses.push(ua.clone());
+
+            assert_eq!(self.addresses.len(), previous_num_addresses + 1);
+            self.addresses_write_lock
+                .swap(false, atomic::Ordering::Release);
+
+            self.transparent_addrs
+                .write()
+                .map(|mut w| w.insert(path_u32, t_pubkey))
+                .map_err(|e| format!("Error: {e}"))?;
+        }
+
+        todo!("Do the same for shielded address, sapling");
+    }
+
+    pub const fn t_derivation_path(coin_type: u32, index: u32) -> [ChildIndex; 5] {
+        [
+            ChildIndex::hardened(44),
+            ChildIndex::hardened(coin_type),
+            ChildIndex::hardened(0),
+            // TODO: Bellow child_index in our lib are meant to be normal
+            // not hardened, but seems zingolib fork of this library
+            // uses only hardened fields in path????
+            ChildIndex::hardened(0),
```

Comment:
This isn't the case, see https://github.com/zingolabs/zingolib/blob/1be3f866fceb12be4e06b04371d783e5bf85e777/zingolib/src/wallet/keys/extended_transparent.rs#L98-L108

The `zcash_primitives::ChildIndex` type being used here only supports hardened indexes because it's a sapling-specific type, and sapling doesn't need non-hardened indexes. We have our own KeyIndex type...that is a holdover from zecwallet-lite, and I think we can use `hdwallet::KeyIndex` in its place, which is the type used by librustzcash
for its transparent derivation